File: 프로그래머스/lv2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    for i in range(m):
        board[i] = list(board[i])
    
    cnt = 0
    rm = set()
    while True:
        for i in range(m-1):
            for j in range(n-1):
                t = board[i][j]
                # 빈칸은 카운트 안함
                if t == []:
                    continue
                if board[i+1][j] == t and board[i][j+1] == t and board[i+1][j+1] == t:
                    rm.add((i,j))
                    rm.add((i+1,j))
                    rm.add((i,j+1))
                    rm.add((i+1,j+1))
        
        # 2*2 블럭이 존재한다면 좌표 개수 세주고 블럭 지우기
        if rm:
            cnt += len(rm)
            print(cnt)
            for i,j in rm:
                board[i][j] = []
            rm = set()
        else: 
            return cnt
                
        # 위에서 내려오면서 블럭 내리기
        while True:
            moved = 0
            for i in range(m-1):
                for j in range(n):
                    if board[i][j] and board[i+1][j] == []:
                        board[i+1][j] = board[i][j]
                        board[i][j] = []
                        moved = 1
            if moved ==0:
                break

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.js
function solution(m, n, board) {
    board = board.map((x) => x.split(""));
    
    while(1) {
        const blocks = [];
        for (let i = 0; i < m - 1; i++) {
            for (let j = 0; j < n - 1; j++) {
                if (board[i][j] && board[i][j] === board[i + 1][j] && board[i][j] === board[i + 1][j + 1] && board[i][j] === board[i][j + 1]) {
                    blocks.push([i, j]);
                }
            }
        }
        
        if (!blocks.length) {
            return [].concat(...board).filter((x) => !x).length;
        }
        
        for (let numbers of blocks) {
            let [i, j] = numbers;
            board[i][j] = 0;
            board[i + 1][j] = 0;
            board[i + 1][j + 1] = 0;
            board[i][j + 1] = 0;
        }
        
        for (let i = m - 1; i > 0; i--) {
            if (!board[i].some((v) => !v)) continue;
            
            for (let j = 0; j < n; j++) {
            const zero = [];
            const text = [];
            
            for (let i = m - 1; i >= 0; i--) {
                !board[i][j] ? zero.push(i) : text.push(i);
            }
            
            if (zero.length !== m) {
                for (let idx in zero) {
                    for (let textIdx in text) {
                        if (zero[idx] > text[textIdx]) {
                            board[zero[idx]][j] = board[text[textIdx]][j];
                            board[text[textIdx]][j] = 0;
                            break;
                            }
                        }
                    }
                }
            }
        }
    }
}
